### PR TITLE
Persist logpoller filters

### PR DESCRIPTION
**Pull request by reductionista**
There is a subtle case where logpoller can miss logs.  The scenario is where a job is added, starts up and properly creates the filter, but then the node is restarted (and possibly reconfigured) causing the job to fail to start.   (For example, if the rpc servers are down or misconfigured after the restart.)  We would like this to be only a temporary problem, where the job can start up and pick up where it left off as soon as the configuration is fixed, or whatever condition is keeping it from starting is fixed.  However, on current develop branch logpoller may not know to follow events for jobs which failed to start, if they failed before the call to RegisterFilter(), leading to a gap where any events which happened during the downtime never get picked up once the job is back up and running.

Instead of trying to enforce that each job always call RegisterFilter before anything else that could fail, we can fix the situation with a more general safety mechanism that should guarantee no missed logs, even after periods of time where existing jobs cannot run.
This is accomplished in this PR by saving all registered filters in a persistent db table instead of only in memory:

- Create log_poller_filters table
- Add FilterName & ChainID fields to Filter struct
- Index in-memory filters by a fixed FilterName instead of temporary integer index which could change after each restart
- Unregister filters by FilterName instead of integer index
- Save each filter to db as soon as it's registered.
- Reload all saved filters when node comes back up after a restart.

### CODE REVIEW

File: core/store/migrate/migrations/0155_add_log_poller_filters_table.sql [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    id BIGSERIAL PRIMARY KEY,
+    filter_name TEXT,
+    address BYTEA CHECK (octet_length(address) = 20) NOT NULL,
+    event BYTEA CHECK (octet_length(event) = 32) NOT NULL,
+    evm_chain_id numeric(78,0) REFERENCES evm_chains (id) DEFERRABLE INITIALLY IMMEDIATE,
+    created_at TIMESTAMPTZ NOT NULL,
+    UNIQUE (filter_name, evm_chain_id, address, event)
+);
+
+CREATE INDEX idx_log_poller_filters_address_event ON log_poller_filters(filter_name);
```
Comment (makramkd) [inline core/store/migrate/migrations/0155_add_log_poller_filters_table.sql:13]: Index name seems misleading - should it be `idx_log_poller_filters_filter_name` instead?

File: core/services/ocr2/plugins/ocr2vrf/coordinator/coordinator.go [hunk 2]
```diff
@@ -169,7 +169,7 @@ func New(
 
 	// Add log filters for the log poller so that it can poll and find the logs that
 	// we need.
-	_, err = logPoller.RegisterFilter(logpoller.Filter{
+	err = logPoller.RegisterFilter(logpoller.Filter{
 		EventSigs: []common.Hash{
```
Comment (makramkd) [inline core/services/ocr2/plugins/ocr2vrf/coordinator/coordinator.go:173]: Is a name supposed to be provided here?

**Comment by makramkd** [comment]
What's the rationale behind persisting the filters?

**Comment by reductionista** [comment]
> What's the rationale behind persisting the filters?

I've updated the description with an explanation of the problem it solves.

**Comment by EasterTheBunny** [comment]
What is the intended usage pattern for filters?

1. Explicitly set and automatically clear (I think this was the old way)
2. Automatically set (from the db after restart) and explicitly clear

I think this is a critical design choice to stay consistent with for long term support and plenty of documentation on describing any magical behavior (loading from the db).

Maybe this behavior could be a configuration option that could be explicitly set? That might help eliminate confusion.

**Comment by reductionista** [comment]
> What is the intended usage pattern for filters?

Each job will continue to be responsible for explicitly calling `RegisterFilter()` and `UnregisterFilter()`.  As before, nothing will be registered or unregistered automatically, the only difference is that if the node goes down with it registered, it will come back up with it still registered.  Calling `RegisterFilter()` more than once is perfectly fine, as both that and `UnregisterFilter()` are intentionally designed to be idempotent.

However, it's true that we're planning for the usage pattern to change a bit, and agreed that this is important to get right.

The main change in usage pattern will be that `UnregisterFilter()` should only be called when the job is getting deleted, rather than during shutdown.  However, out of a dozen or so clients of logpoller, only one (ocr2keeper registry) currently bothers to make the call to `UnregisterFilter()`.  Aside from that one call in the codebase, `UnregisterFilter()` was entirely unused; so it's more like a new step in the recommended flow being added rather than an existing one being changed.

Ability for individual jobs to do this in an atomic way will be added soon in a planned followup PR, as well as the actual calls to `UnregisterFilter()`.  Some of the details are still being worked out, but tentatively what we have in mind is for it to happen in a new method of the job interface, probably called `OnDeleteJob()`.  The call to it will have to be passed down through the plugins/relay/provider stack.  Currently we have `BeforeCreateJob()`, `AfterCreateJob()` and `BeforeDeleteJob()`.   Unlike `BeforeDeleteJob()`, `OnDeleteJob()` will be from inside the same db tx which contains the DELETE sql command for the job, ensuring that if anything goes wrong they can both be rolled back atomically.

A spike on this, as well as the possibility of moving the calls to `RegisterFilter()` into `BeforeCreateJob()` will begin as soon as this PR gets merged.  In the meantime, I'm removing the single call to `UnregisterFilter()`, although I suppose it would also work if we kept it.  Either way, we plan to have the calls to `UnregisterFilter()` in place before anything is officially released.

File: core/chains/evm/logpoller/log_poller.go [hunk 3]
```diff
@@ -260,6 +369,16 @@ func (lp *logPoller) Start(parentCtx context.Context) error {
 		ctx, cancel := context.WithCancel(parentCtx)
 		lp.ctx = ctx
 		lp.cancel = cancel
+		filters, err := lp.orm.LoadFilters(pg.WithParentCtx(lp.ctx))
+		lp.filterDirty = true
+		if err != nil {
+			// Most likely, we will get them anyway once RegisterFilter() is called, but
```
Comment (krehermann) [inline core/chains/evm/logpoller/log_poller.go:375]: the comment confuses me.

why wouldn't start fail if loading the filters fails? is as, can it lead to an inconsistent state if there an error?

File: core/chains/evm/logpoller/log_poller_test.go [hunk 4]
```diff
@@ -418,8 +419,30 @@ func TestLogPoller_Logs(t *testing.T) {
...
+			fmt.Sprintf("db filter table %s is missing some addresses or events from in-memory filter", name))
+	}
+}
+
 func TestLogPoller_RegisterFilter(t *testing.T) {
-	lp := NewLogPoller(nil, nil, nil, 15*time.Second, 1, 1, 2, 1000)
+	lggr := logger.TestLogger(t)
+	chainID := testutils.NewRandomEVMChainID()
+	db := pgtest.NewSqlxDB(t)
+	require.NoError(t, utils.JustError(db.Exec(`SET CONSTRAINTS log_poller_filters_evm_chain_id_fkey DEFERRED`)))
```
Comment (krehermann) [inline core/chains/evm/logpoller/log_poller_test.go:442]: out of curiosity - what is the intention of this db statement? why is it needed? it seems magical, maybe consider a comment

File: core/chains/evm/logpoller/orm.go [hunk 5]
```diff
@@ -37,6 +37,54 @@ func (o *ORM) InsertBlock(h common.Hash, n int64, qopts ...pg.QOpt) error {
...
+// InsertFilter is idempotent.
+//
+// Each address/event pair must have a unique job id, so it may be removed when the job is deleted.
+// If a second job tries to overwrite the same pair, this should fail.
+func (o *ORM) InsertFilter(filter Filter, qopts ...pg.QOpt) (err error) {
+	q := o.q.WithOpts(qopts...)
+	addresses := make([][]byte, 0)
+	events := make([][]byte, 0)
+
+	for _, addr := range filter.Addresses {
```
Comment (krehermann) [inline core/chains/evm/logpoller/orm.go:49]: do you need a read lock on the filter?

File: core/services/ocr2/plugins/ocr2keeper/log_provider.go [hunk 6]
```diff
@@ -34,15 +35,17 @@ func NewLogProvider(
 	lookbackBlocks int64,
 ) (*LogProvider, error) {
 	var err error
+	filterName := fmt.Sprintf("OCR2KeeperRegistry - %s", registryAddress)
```
Comment (krehermann) [inline core/services/ocr2/plugins/ocr2keeper/log_provider.go:38]: i can't tell from the code -- are these names persisted to the db? would it be problematic if they change from one release/version to another?

File: core/chains/evm/logpoller/log_poller_test.go (hunk 4 above)
Comment (reductionista) [reply core/chains/evm/logpoller/log_poller_test.go:442]: It's telling the db to ignore any constraint violations due to the chain id not being found in the `evm_chains` table.  I noticed that's how it's done inside `SetupTH` for two similar constraints, so just copied the same pattern:
https://github.com/smartcontractkit/chainlink/pull/8427/files#diff-05e070919c02c8a135a89fbe988cbc3f4a83736f57c98fd6290b162edef08e52L44-L45

Usually, DEFERRED wouldn't imply ignored, but I think the reason it works in this case is due to the way our unit tests are set up:  all tests are encapsulated inside a single transaction, which gets rolled back at the end... but only after all the tests have either passed or failed.

File: core/chains/evm/logpoller/log_poller.go (hunk 3 above)
Comment (reductionista) [reply core/chains/evm/logpoller/log_poller.go:375]: It could fail due to a timeout if there is a temporary network outage, or the db or the node is too overloaded with work and slow at processing requests.  Or maybe even some subtle error in the SQL statement or decoding/encoding of params, perhaps that depends on the version of postgres.  Later, when the job starts it will try to add the filter, but the job itself constructs the filter so that doesn't depend on the db.  Some of those could be indications of more widespread problem that would cause the job to be unable to start.  Perhaps "most likely" is overly optimistic, but either way it seems best to assume it will start okay instead of refuse to start logpoller which would in itself prevent the job from being able to run properly.

You could consider it an inconsistent state in that before the job starts, the filter would be in db but not in memory, but that would be fixed as soon as the job starts.  In fact, that would be no different from current behavior... we start logpoller with no filters in memory, then let the individual job services add them when they start.

File: core/services/ocr2/plugins/ocr2keeper/log_provider.go (hunk 6 above)
Comment (reductionista) [reply core/services/ocr2/plugins/ocr2keeper/log_provider.go:38]: Yes, they are persisted to db.  I think in generally we don't consider the db tables to be "user facing", unless things are read from there and displayed in the UI or console.  These strings would be in the tables, and could show up in logs.  We could still change them with an appropriate db migration, so I think it's okay.

We could make all of these random integers to avoid having to migrate the db if they change, but... I think I'd prefer the logs to be more readable, in case of any problems come up that require a rapid response.

File: core/chains/evm/logpoller/log_poller.go [hunk 7]
```diff
@@ -116,14 +117,104 @@ func NewLogPoller(orm *ORM, ec Client, lggr logger.Logger, pollPeriod time.Durat
...
+	for i, ba := range baArray.Elements {
+		addr := common.Address{}
+		if ba.Status != pgtype.Present {
+			return errors.Errorf("Expected all addresses in AddressArray to be non-NULL.  Got AddressArray[%d] = NULL", i)
+		}
+		err = addr.Scan(ba.Bytes)
+		if err != nil {
+			return err
+		}
+		if addr != lastAddress { // dedup addresses
```
[the remark this answers is not in the document]
Comment (reductionista) [reply core/chains/evm/logpoller/log_poller.go:156]: I intended to guarantee they would be ordered, but look like there's a bug in my sql query... they have the wrong ORDER BY clause.  Thanks!

[the remark this answers is not in the document]
Comment (reductionista) [reply core/chains/evm/logpoller/log_poller.go:156]: Actually, I think this would all be a lot cleaner if I changed the db schema to have an array of events & topics on each row, rather than flattening them out to have one address,topic pair on each row.  Then there would be no need for deduping.

The only reason I avoided that, is in case in the future we might want to search by contract address or topic.  But maybe that's unlikely?

[the remark this answers is not in the document]
Comment (reductionista) [reply core/chains/evm/logpoller/log_poller.go:156]: Oh!  Found a better solution:  apparently you can use DISTINCT inside ARRAY_AGG; that removes the need for deduping in Scan()

File: core/services/ocr2/plugins/ocr2keeper/log_provider.go (hunk 6 above)
Comment (jmank88) [reply core/services/ocr2/plugins/ocr2keeper/log_provider.go:38]: Would it make sense to provide a helper func or type to format these all in the same standard way, rather than relying on callers to follow the pattern? Would keep things aligned and make it simpler to refactor later if necessary too. For example:
```go
logpoller.FilterName("OCR2KeeperRegistry", registryAddress)

func FilterName(id string, args ...any) string {
	var s strings.Builder
	s.WriteString(id)
	s.WriteString(" - ")
	fmt.Fprintf("%s", args[0])
	for _, a := range args[1:] {
		fmt.Fprintf(":%s", a)
	}
	return s.String()
}
```

File: core/store/migrate/migrations/0155_add_log_poller_filters_table.sql [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    id BIGSERIAL PRIMARY KEY,
+    filter_name TEXT,
+    address BYTEA CHECK (octet_length(address) = 20) NOT NULL,
+    event BYTEA CHECK (octet_length(event) = 32) NOT NULL,
+    evm_chain_id numeric(78,0) REFERENCES evm_chains (id) DEFERRABLE INITIALLY IMMEDIATE,
+    created_at TIMESTAMPTZ NOT NULL,
+    UNIQUE (filter_name, evm_chain_id, address, event)
+);
+
+CREATE INDEX idx_log_poller_filters_filter_name ON log_poller_filters(filter_name);
```
Comment (connorwstein) [inline core/store/migrate/migrations/0155_add_log_poller_filters_table.sql:13]: I think this is not needed UNIQUE constraint automatically creates a multi-column index

File: core/chains/evm/logpoller/log_poller.go [hunk 9]
```diff
@@ -104,7 +105,8 @@ type ReplayRequest struct {
 // support chain, polygon, which has 2s block times, we need RPCs roughly with <= 500ms latency
 func NewLogPoller(orm *ORM, ec Client, lggr logger.Logger, pollPeriod time.Duration,
 	finalityDepth int64, backfillBatchSize int64, rpcBatchSize int64, keepBlocksDepth int64) *logPoller {
-	return &logPoller{
```
Comment (connorwstein) [inline core/chains/evm/logpoller/log_poller.go:107]: why this change? generally go advises against defining a short var like x := &s since it masks the pointer nature of the var hurting readability (usually preferable to & at stack frame boundaries to make it clear its returning a pointer)

File: core/chains/evm/logpoller/log_poller.go (hunk 3 above)
Comment (connorwstein) [reply core/chains/evm/logpoller/log_poller.go:330]: hmm I'd treat this the same as SelectLatestBlock, its just DB state we want to load before starting to poll. We can keep retrying forever in the log poll tick loop and only start progressing from last saved block once we've successfully loaded the filters. Then we can cleanly recover from transient DB issues and never miss any logs

Comment (connorwstein) [reply core/chains/evm/logpoller/log_poller.go:330]: Kind of a general pattern for synchronizing external state I've seen a bunch. 1) On boot try forever to load initial state and a progress identifier (i.e. block pointer) from DB 2) While up atomically update state and block pointer in a DB tx. Doesn't matter if the node or DB crashes/flaps up and down etc. you never miss anything or have any partial updates.

File: core/chains/evm/logpoller/log_poller.go (hunk 9 above)
Comment (reductionista) [reply core/chains/evm/logpoller/log_poller.go:107]: Not intentional, I probably changed it while debugging or refactoring, and forgot to put it back.  Hadn't heard that about preferring & at stack boundary thanks for the tip.

File: core/chains/evm/logpoller/log_poller.go [hunk 10]
```diff
@@ -116,14 +118,56 @@ func NewLogPoller(orm *ORM, ec Client, lggr logger.Logger, pollPeriod time.Durat
...
+	Addresses  evmtypes.AddressArray
+}
+
+// FilterName is a suggested convenience function for clients to construct unique filter names
+// to populate FilterName field of struct Filter
+func FilterName(id string, args ...any) string {
+	s := &strings.Builder{}
+	s.WriteString(id)
+	s.WriteString(" - ")
+	fmt.Fprintf(s, "%s", args[0])
```
[the remark this answers is not in the document]
Comment (reductionista) [reply core/chains/evm/logpoller/log_poller.go:140]: I added a couple unit tests for it last night, might have gone in while your review was pending.  Good catch on the empty args

File: core/chains/evm/logpoller/orm.go [hunk 11]
```diff
@@ -37,6 +37,53 @@ func (o *ORM) InsertBlock(h common.Hash, n int64, qopts ...pg.QOpt) error {
...
+	q := o.q.WithOpts(qopts...)
+	err := q.ExecQ(`DELETE FROM log_poller_filters WHERE filter_name = $1`, name)
+	return err
+}
+
+// LoadFiltersForChain returns all filters for this chain
+func (o *ORM) LoadFilters(qopts ...pg.QOpt) (map[string]Filter, error) {
+	q := o.q.WithOpts(qopts...)
+	rows := make([]Filter, 0)
+	err := q.Select(&rows, `SELECT filter_name, ARRAY_AGG(DISTINCT address)::BYTEA[] AS addresses, ARRAY_AGG(DISTINCT event)::BYTEA[] AS event_sigs
```
[the remark this answers is not in the document]
Comment (reductionista) [reply core/chains/evm/logpoller/orm.go]: Yes, although it's not very specific to this query.  We will need them any time we want to read data directly into a `Filter` struct, or any other struct which includes arrays of `common.Address` or `common.Hash`.  The underlying reason is because `pgx` doesn't know how to automatically convert from `[]BYTEA[]` to `[20][]byte`, it only works with a data type that looks like `[][]byte`.  Fixed length types and variable length types are handled very differently in sql, and unfortunately there is no fixed length type that looks like `BYTEA[20]`, there is only `BYTEA[]`.  We enforce the 20 with a constraint, but it's not constraint-aware enough to understand that.

**Review by connorwstein — APPROVED** [review]
